gateway/shard/manager.go

```diff
@@ -0,0 +1,314 @@
+	id.IdentifyGlobalLimit.SetBurst(botData.StartLimit.Remaining)
+	resetAt := time.Now().Add(botData.StartLimit.ResetAfter.Duration())
+	id.IdentifyGlobalLimit.SetBurstAt(resetAt, botData.StartLimit.Total)
+
+	// Update the maximum number of identify requests allowed per 5s.
+	id.IdentifyShortLimit.SetBurst(botData.StartLimit.MaxConcurrency)
+
+	return botData.URL, nil
+}
+
+// Manager is the manager responsible for handling all sharding on this
+// instance. An instance of Manager must never be copied.
+type Manager struct {
+	// shards are the *shards.shards managed by this Manager. They are
+	// sorted in ascending order by their shard id.
+	shards     []shard
+	gatewayURL string
+
+	mutex sync.RWMutex
+
+	rescaling *rescalingState // nil unless rescaling
+
+	new NewShardFunc
+}
+
+type rescalingState struct {
+	haltRescale context.CancelFunc
+	rescaleDone sync.WaitGroup
+}
+
+// NewManager creates a Manager using as many gateways as recommended by
+// Discord.
+func NewManager(token string, new NewShardFunc) (*Manager, error) {
+	id := gateway.DefaultIdentifier(token)
+
+	url, err := updateIdentifier(context.Background(), id)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get gateway info")
+	}
+
+	return NewIdentifiedManagerWithURL(url, id, new)
+}
+
+// NewIdentifiedManager creates a new Manager using the given
+// gateway.Identifier. The total number of shards will be taken from the
+// identifier instead of being queried from Discord, but the shard ID will be
+// ignored.
+//
+// This function should rarely be used, since the shard information will be
+// queried from Discord if it's required to shard anyway.
+func NewIdentifiedManager(data gateway.IdentifyData, new NewShardFunc) (*Manager, error) {
+	// Ensure id.Shard is never nil.
+	if data.Shard == nil {
+		data.Shard = gateway.DefaultShard
+	}
+
+	id := gateway.NewIdentifier(data)
+
+	url, err := updateIdentifier(context.Background(), id)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get gateway info")
+	}
+
+	return NewIdentifiedManagerWithURL(url, id, new)
+}
+
+// NewIdentifiedManagerWithURL creates a new Manager with the given Identifier
+// and gateway URL. It behaves similarly to NewIdentifiedManager.
+func NewIdentifiedManagerWithURL(
+	url string, id *gateway.Identifier, new NewShardFunc) (*Manager, error) {
+
+	m := Manager{
+		gatewayURL: gateway.AddGatewayParams(url),
+		shards:     make([]shard, id.Shard.NumShards()),
+		new:        new,
+	}
+
+	var err error
+
+	for i := range m.shards {
+		data := id.IdentifyData
+		data.Shard = &gateway.Shard{i, len(m.shards)}
+
+		m.shards[i] = shard{
+			id: gateway.Identifier{
+				IdentifyData:        data,
+				IdentifyShortLimit:  id.IdentifyShortLimit,
+				IdentifyGlobalLimit: id.IdentifyGlobalLimit,
+			},
+		}
+
+		m.shards[i].Shard, err = new(&m, &m.shards[i].id)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create shard %d/%d", i, len(m.shards)-1)
+		}
+	}
+
+	return &m, nil
+}
```

Comment:
`new` should be replaced here as well.
```suggestion
func NewIdentifiedManagerWithURL(
	url string, id *gateway.Identifier, fn NewShardFunc) (*Manager, error) {

	m := Manager{
		gatewayURL: gateway.AddGatewayParams(url),
		shards:     make([]shard, id.Shard.NumShards()),
		new:        fn,
	}

	var err error

	for i := range m.shards {
		data := id.IdentifyData
		data.Shard = &gateway.Shard{i, len(m.shards)}

		m.shards[i] = shard{
			id: gateway.Identifier{
				IdentifyData:        data,
				IdentifyShortLimit:  id.IdentifyShortLimit,
				IdentifyGlobalLimit: id.IdentifyGlobalLimit,
			},
		}

		m.shards[i].Shard, err = fn(&m, &m.shards[i].id)
		if err != nil {
			return nil, errors.Wrapf(err, "failed to create shard %d/%d", i, len(m.shards)-1)
		}
	}

	return &m, nil
}
```